README.md
# Print Friendly Plugin

![Print Friendly](assets/printfriendly.png)

`Print Friendly` is a [Grav](http://github.com/getgrav/grav) Plugin that generates a version of a [web page formatted for printing](https://en.wikipedia.org/wiki/Printer-friendly).

# Installation

Installing the `Print Friendly` plugin can be done in one of two ways. The GPM (Grav Package Manager) installation method enables you to quickly and easily install the plugin with a simple terminal command, while the manual method enables you to do so via a zip file.

## GPM Installation (Preferred)

The simplest way to install this plugin is via the [Grav Package Manager (GPM)](http://learn.getgrav.org/advanced/grav-gpm) through your system's Terminal (also called the command line).  From the root of your Grav install type:

    bin/gpm install printfriendly

This will install the `Print Friendly` plugin into your `/user/plugins` directory within Grav. Its files can be found under `/your/site/grav/user/plugins/printfriendly`.

## Manual Installation

To install this plugin, just download the zip version of this repository and unzip it under `/your/site/grav/user/plugins`. Then, rename the folder to `printfriendly`. You can find these files either on [GitHub](https://github.com/iusvar/grav-plugin-printfriendly) or via [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

You should now have all the plugin files under

    /your/site/grav/user/plugins/printfriendly

# Updating

As development for the `Print Friendly` plugin continues, new versions may become available that add additional features and functionality, improve compatibility with newer Grav releases, and generally provide a better user experience. Updating `Print Friendly` is easy, and can be done through Grav's GPM system, as well as manually.

## GPM Update (Preferred)

The simplest way to update this plugin is via the Grav Package Manager (GPM). You can do this with this by navigating to the root directory of your Grav install using your system's Terminal (also called command line) and typing the following:

    bin/gpm update printfriendly

This command will check your Grav install to see if your `Print Friendly` plugin is due for an update. If a newer release is found, you will be asked whether or not you wish to update. To continue, type y and hit enter. The plugin will automatically update and clear Grav's cache.

## Manual Update

Manually updating `Print Friendly` is pretty simple. Here is what you will need to do to get this done:

* Delete the `your/site/user/plugins/printfriendly` directory.

* Downalod the new version of the Print Friendly plugin from either [GitHub](https://github.com/iusvar/grav-plugin-printfriendly) or [GetGrav.org](http://getgrav.org/downloads/plugins#extras).

* Unzip the zip file in `your/site/user/plugins` and rename the resulting folder to `printfriendly`.

* Clear the Grav cache. The simplest way to do this is by going to the root Grav directory in terminal and typing `bin/grav clear-cache`.

Note: Any changes you have made to any of the files listed under this directory will also be removed and replaced by the new set. Any files located elsewhere (for example a YAML settings file placed in `user/config/plugins`) will remain intact.

# Usage

The plugin was tested in the following updated environments:

* Grav v1.5.3
* Admin v1.8.12
* Quark v1.2.2
* Antimatter v2.1.2
* Learn2 v1.6.3
* Knowledge Base v1.3.3

For now the `printfriendly` plugin as been tested for four types of sites:

* [Blog Site](https://github.com/getgrav/grav-skeleton-blog-site) skeleton with [Quark](https://github.com/getgrav/grav-theme-quark) theme by [Grav Team](https://getgrav.org/about)
* [Blog Site](https://github.com/getgrav/grav-skeleton-blog-site) skeleton with [Antimatter](https://github.com/getgrav/grav-theme-antimatter) theme by [Grav Team](https://getgrav.org/about)
* [RTFM Site](https://github.com/getgrav/grav-skeleton-rtfm-site) skeleton with [Learn2](https://github.com/getgrav/grav-theme-learn2) theme by [Grav Team](https://getgrav.org/about)
* [Knowledge Base](https://github.com/Perlkonig/grav-skeleton-knowledge-base) skeleton with [Knowledge Base](https://github.com/Perlkonig/grav-theme-knowledge-base) theme by [Aaron Dalton](https://github.com/Perlkonig)

To use this plugin, you simplest way to display the link to print is to use the Twig function and pass it a dynamic value such as:

```
{{ printfriendly(page.route) }}
```
or, in case of Autoescape variables set as yes in the system.yaml,
```
{{ printfriendly(page.route)|raw }}
```

This will use the current page's route as the unique identifier.

The `Print Friendly` plugin comes with some sensible default configuration, that are pretty self explanatory:

## Config Defaults

```yaml
# Global plugin configurations.
enabled: true                   # Set to false to disable this plugin completely.
built_in_css: true              # Use built-in CSS of the plugin.

basics:
  btn_plugin: 
  icn_plugin: 
  print_directly: false         # Print directly avoiding the window. In this case the library JQuery UI is not needed.

# jQuery user interface library, Font Awesome, the iconic font and CSS framework.
libraries:
  jqueryui_source: maxcdn       # Set source between your server or MaxCDN.
  jqueryui_version: 1.12.0      # Set version.
  jqueryui_themes: start        # Set theme for window.
  awesome_use_font: false       # Set true to use the library.

# Some items you can include in printing.
settings:

  page:
    image: true                 # It refers to the first image in the folder.
                                # However, all images including the first are printed if the img tag is present in allowed_tags.
    title: true
    author: true
    date: true
    date_format: short
    tags: true

  breadcrumbs:
    enabled: true
    link: true
    current: true

# The strip_tags function delete from a html document all tags except those listed.
# You can add or delete tags elements.
tags:
  allowed_tags:
    - a
    - b
    - blockquote
    - br
    - center
    - cite
    - code
    - div
    - em
    - h1
    - h2
    - h3
    - h4
    - h5
    - h6
    - hr
    - i
    - img
    - li
    - ol
    - p
    - pre
    - small
    - span
    - strong
    - sub
    - sup
    - table
    - tbody
    - td
    - textarea
    - tfoot
    - th
    - thead
    - tr
    - u
    - ul

# Window size.
window:
  width: 50                     # A percentage of the maximum size of the device width.
  height: 40                    # A percentage of the maximum size of the device height.
  closeonescape: false
  close_icon: false
  confirm_button: Print
  cancel_button: Close
```
If you need to change any value, then the best process is to copy the [printfriendly.yaml](printfriendly.yaml) file into your `users/config/plugins/` folder (create if doesn't exist), and modify it there. This will override the default settings.


CHANGELOG.md
# v0.7.2
## 11/05/2018

1. [](#bugfix)
    * Add raw filter in the twig function.
    
# v0.7.1
## 11/05/2018

1. [](#bugfix)
    * [#9](https://github.com/iusvar/grav-plugin-printfriendly/issues/9) Fix the reserved indicator "%" that cannot start a plain scalar.
    * Make the plugin to work with Twig auto-escaping turned on and then with the Quark theme.
    
# v0.7.0
## 05/17/2018

1. [](#improved)
    * Restored coding with base64.

# v0.6.0
## 05/15/2018

1. [](#bugfix)
    * Fix characters that do not appear in Russian [#8](https://github.com/iusvar/grav-plugin-printfriendly/issues/8) and Turkish [#7](https://github.com/iusvar/grav-plugin-printfriendly/issues/7) language. 

# v0.5.0
## 08/03/2017

1. [](#improved)
    * Added parameter param_sep in url construction.
    * New site with working [demo](http://iusvar.alwaysdata.net/grav/). [Snappygrav](https://github.com/iusvar/grav-plugin-snappygrav)'s demo also works: On the Alwaysdata hosting server, wkhtmltopdf works wonderfully.

# v0.4.0
## 07/13/2017

1. [](#new)
    * Newly added nonce features also to identify html div.
1. [](#bugfix)
    * The data to be printed is created only when requested

# v0.3.3
## 07/10/2017

1. [](#bugfix)
    * Simplified procedure
    * Eliminated Nonce features
    * Maybe we're with [#2](https://github.com/iusvar/grav-plugin-printfriendly/issues/2) and [#4](https://github.com/iusvar/grav-plugin-printfriendly/issues/4) bugs.

# v0.3.2
## 07/07/2017

1. [](#bugfix)
    * Fix right version
    * Improved slug selection. It may be the cause of the [#4](https://github.com/iusvar/grav-plugin-printfriendly/issues/4) error.

# v0.3.1
## 07/06/2017

1. [](#bugfix)
    * Deleted event onCollectionProcessed: only for debugging
    * Fix Twig_Error_Runtime: An exception has been thrown during the rendering of a template ("Only variables should be passed by reference").[#3](https://github.com/iusvar/grav-plugin-printfriendly/issues/3)

# v0.3.0
## 07/06/2017

1. [](#new)
    * Added Nonce features for printing
    * Added pf_implode Twig filter
    * Added configuration option to get breadcrumbs
    * Added configuration option to get tags
    * Added configuration option to get first image
    * Added configuration option to set Close On Escape
    * Added configuration option to set Close Icon
    * Added configuration option to set Confirm and Cancel Buttons Text
    * Added animated icon
    * Added icons on the buttons
1. [](#improved)
    * Improved printfriendly Twig function
    * Updated jQuery UI library to v1.12.1
    * Updated Font Awesome library to v4.7.0
    * English and Italian languages improvements
    * Deleted superflue javascript library
1. [](#bugfix)
    * Fix Broken internal anchors [#2](https://github.com/iusvar/grav-plugin-printfriendly/issues/2). All information to be printed is entered in a hidden input field

# v0.2.1
## 08/29/2016

1. [](#improved)
    * Print directly avoiding the window

# v0.1.1
## 08/23/2016

1. [](#improved)
    * Added french (@MATsxm) language
1. [](#bugfix)
    * Correct README.md link

# v0.1.0
##  08/23/2016

1. [](#new)
    * Initial release.


classes/pfmanager.php
<?php

namespace Grav\Plugin\PFManager;
use Grav\Common\Grav;
use Grav\Common\Utils;
use RocketTheme\Toolbox\File\JsonFile;

class PFManager
{
    public $grav;
    public $json_response;
    protected $post;
    protected $task;

    protected $lang;

    public function __construct(Grav $grav)
    {
        $this->grav     = $grav;
        $this->config   = $this->grav['config'];
        $this->lang     = $this->grav['language'];
    }


    public function setMessage($msg, $type = 'info')
    {
        $messages = $this->grav['messages'];
        $messages->add($msg, $type);
    }


    public function messages($type = null)
    {
        $messages = $this->grav['messages'];
        return $messages->fetch($type);
    }


    public function execute($task, $post)
    {
        $this->task = $task;
        $this->post = $post;
        if (!$this->validateNonce()) {
            return false;
        }

        $method = 'task' . ucfirst($this->task);

        if (method_exists($this, $method)) {
            try {
                $success = call_user_func([$this, $method]);
            } catch (\RuntimeException $e) {
                $success = true;
                $this->setMessage($e->getMessage(), 'error');
            }
        }

        return $success;
    }


    protected function validateNonce()
    {
        if (method_exists('Grav\Common\Utils', 'getNonce')) {
            if (strtolower($_SERVER['REQUEST_METHOD']) == 'post') {
                if (isset($this->post['pf-nonce'])) {
                    $nonce = $this->post['pf-nonce'];
                } else {
                    $nonce = $this->grav['uri']->param('pf-nonce');
                }

                if (!$nonce || !Utils::verifyNonce($nonce, 'pf-form')) {
                    if ($this->task == 'addmedia') {

                        $message = sprintf($this->lang->translate('PLUGIN_ADMIN.FILE_TOO_LARGE', null),
                            ini_get('post_max_size'));

                        $this->json_response = [
                            'status'  => 'error',
                            'message' => $message
                        ];
                        return false;
                    }

                    $this->setMessage($this->lang->translate('PLUGIN_ADMIN.INVALID_SECURITY_TOKEN'), 'error');
                    $this->json_response = [
                        'status'  => 'error',
                        'title' => $this->lang->translate('PLUGIN_PF.INVALID_SECURITY_TOKEN_TITLE'),
                        'message' => $this->lang->translate('PLUGIN_PF.INVALID_SECURITY_TOKEN_TEXT')
                    ];
                    return false;
                }
                unset($this->post['pf-nonce']);
            } else {
                $nonce = $this->grav['uri']->param('pf-nonce');
                if (!isset($nonce) || !Utils::verifyNonce($nonce, 'pf-form')) {
                    $this->setMessage($this->lang->translate('PLUGIN_PF.INVALID_SECURITY_TOKEN'), 'error');
                    $this->json_response = [
                        'status'  => 'error',
                        'title' => $this->lang->translate('PLUGIN_PF.INVALID_SECURITY_TOKEN_TITLE'),
                        'message' => $this->lang->translate('PLUGIN_PF.INVALID_SECURITY_TOKEN_TEXT')
                    ];
                    return false;
                }
            }
        }
        return true;
    }


    protected function taskPf()
    {
        $route  = $this->grav['uri']->param('route');
        $route  = str_replace('@','/',$route);

        try {
            $ret_value  = $this->makeDocument($route);
            $title = $ret_value['title'];
            $html_base64_encode = $ret_value['html_base64_encode'];
        } catch (\Exception $e) {
            $this->json_response = [
                'status' => 'error',
                'title' => 'taskPf Error',
                'message' => $this->lang->translate('PLUGIN_PF.AN_ERROR_OCCURRED') . '. ' . $e->getMessage()
            ];
            return true;
        }

        $btn_plugin = $this->lang->translate('PLUGIN_PF.BTN_PLUGIN');
        $message    = $this->lang->translate('PLUGIN_PF.YOUR_DOCUMENT_IS_READY_FOR');

        $this->json_response = [
            'status' => 'success',
            'message' => $message,
            'title' => $title,
            'html_base64_encode' => $html_base64_encode
        ];

        return true;
    }


    protected function makeDocument($route)
    {
        $page   = $this->grav['page'];
        $found  = $page->find( $route);
        $title  = $found->title();

        $parameters = [];
        $parameters['breadcrumbs'] = $this->get_crumbs( $found );

        $twig = $this->grav['twig'];
        $html_from_template = $twig->processTemplate('printfriendly.html.twig', ['page' => $found, 'parameters' => $parameters]);

        $allow_array = $this->config->get('plugins.printfriendly.tags.allowed_tags');
        $allow = '';
        foreach ($allow_array as $key => $value) {
            if($value){
                $allow .= '<'.$value.'>';
            }
        }
        $stripped_html = strip_tags( $html_from_template, $allow );
        $html_base64_encode = base64_encode( $stripped_html );

        $ret_value = array();
        $ret_value['title'] = $title;
        $ret_value['html_base64_encode'] = $html_base64_encode;

        return $ret_value;
    }


    protected function get_crumbs( $page )
    {
        $current = $page;
        $hierarchy = array();
        while ($current && !$current->root()) {
            $hierarchy[$current->url()] = $current;
            $current = $current->parent();
        }
        $home = $this->grav['pages']->dispatch('/');
        if ($home && !array_key_exists($home->url(), $hierarchy)) {
            $hierarchy[] = $home;
        }
        $elements = array_reverse($hierarchy);
        $crumbs = array();
        foreach ($elements as $key => $crumb) {
            $crumbs[] = [ 'route' => $crumb->route(), 'title' => $crumb->title() ];
        }

        return $crumbs;
    }

}


assets/js/printfriendly.js
/*
 * TODO:
 *
 *
 * DONE:
 * labels translation: title, text, close
 * allow to change from admin, width and height of the window
 * awesome icon selected by admin, in front of the title
 * beginning scrolling
 * Print directly avoiding the window. In this case the library JQuery UI is not needed.
 *
 */
function onPrint(title, slug, print, close, width, height, icon, print_directly) {
    if(print_directly){
        $("#printContent-"+slug).printThis();
    } else {
        var w = screen.width*width/100;
        var h = screen.height*height/100
        w = w.toFixed();
        h = h.toFixed();
        //console.log(w);
        //console.log(h);
        $("#dialogDiv-"+slug).dialog({
            width: w,
            height: h,
            title: title,
            modal: true,
            closeOnEscape: false,
            buttons: [{
                tabIndex: -1,
                text: print,
                click: function () {
                    //console.log( $( this ).parents(".ui-dialog") );
                    $("#printContent-"+slug).printThis({
                        appendToEl: $( this ).parents( ".ui-dialog" )
                    });
                    //$('#printContent-'+slug).printThis();
                    $(this).dialog("close");
                    $(this).dialog("destroy");
                }
            }, {
                tabIndex: -1,
                text: close,
                click: function () {
                    $(this).dialog("close");
                    $(this).dialog("destroy");
                }
            }],
            open: function (event, ui) {
                $(".ui-dialog-titlebar-close").hide();
                $( ".ui-dialog-title" ).before( "<i class=\"fa fa-"+icon+"\" aria-hidden=\"true\"></i>&nbsp;" );
                $( ".ui-dialog-title" ).css('float','none');
                $( this ).scrollTop(0);
            }
        }, "open");
        event.preventDefault();
    }
}




printfriendly.php
<?php
namespace Grav\Plugin;

use Grav\Common\Plugin;
use RocketTheme\Toolbox\Event\Event;
use Grav\Common\Utils;
use Symfony\Component\Yaml\Yaml;
use Grav\Plugin\PFManager\PFManager;


/**
 * Class PrintFriendlyPlugin
 * @package Grav\Plugin
 */
class PrintFriendlyPlugin extends Plugin
{
    protected $lang;
    protected $uri;
    protected $pfmanager;
    protected $json_response = [];

    protected $listing = array();

/**
     * @return array
     *
     * The getSubscribedEvents() gives the core a list of events
     *     that the plugin wants to listen to. The key of each
     *     array section is the event that the plugin listens to
     *     and the value (in the form of an array) contains the
     *     callable (or function) as well as the priority. The
     *     higher the number the higher the priority.
     */
    public static function getSubscribedEvents() {
        return [
            'onPluginsInitialized' => [['setting', 1000],['onPluginsInitialized', 0]],
        ];
    }

    public function setting()
    {
        require_once __DIR__ . '/classes/pfmanager.php';
        
        $this->lang = $this->grav['language'];
        $this->uri = $this->grav['uri'];
    }

    /**
     * Activate plugin if path matches to the configured one.
     */
    public function onPluginsInitialized()
    {
        // Don't proceed if we are in the admin plugin
        if ($this->isAdmin()) {
            return;
        }

        // Enable the main event we are interested in
        $this->enable([
            'onTwigInitialized'     => ['onTwigInitialized', 0],
            'onPageInitialized'     => ['onPageInitialized', 1001],
            'onTwigSiteVariables'   => ['onTwigSiteVariables', 0],
            'onTwigTemplatePaths'   => ['onTwigTemplatePaths', 0]
        ]);

        $this->pfmanager = new PFManager($this->grav);
        $this->pfmanager->json_response = [];
        $this->grav['pfmanager'] = $this->pfmanager;
    }


    /**
     * Add current directory to Twig lookup paths.
     */
    public function onTwigTemplatePaths()
    {
        $this->grav['twig']->twig_paths[] = __DIR__ . '/templates';
    }


    public function onPageInitialized()
    {
        $post = !empty($_POST) ? $_POST : [];

        $task = !empty($post['pftask']) ? $post['pftask'] : $this->uri->param('pftask');

        if ($task) {
            $this->pfmanager->execute($task, $post);
            if ( $this->pfmanager->json_response ) {
                echo json_encode($this->pfmanager->json_response);
                exit();
            }
        } else {
        }
    }

    /**
     * Add `printfriendly()` Twig function
     * Add `pf_implode` Twig filter
     */
    public function onTwigInitialized()
    {
        $this->grav['twig']->twig()->addFunction(
            new \Twig_SimpleFunction('printfriendly', [$this, 'generateLink'])
        );

        $this->grav['twig']->twig()->addFilter(
            new \Twig_SimpleFilter('pf_implode', 'implode')
        );
    }

    /**
     * Add CSS and JS to page header
     */
    public function onTwigSiteVariables()
    {
        if ($this->config->get('plugins.printfriendly.built_in_css')) {
            $this->grav['assets']->addCss('plugins://printfriendly/assets/css/printfriendly.css');
        }

        if ($this->config->get('plugins.printfriendly.libraries.jqueryui_version')){
            $version = (string)$this->config->get('plugins.printfriendly.libraries.jqueryui_version');
        } else {
            $version = '1.12.1';
        }

        if ( $this->config->get('plugins.printfriendly.libraries.jqueryui_source') == 'maxcdn' ) {
            $themes_source = 'https://code.jquery.com/ui/'.$version.'/themes/';
            $jquery_ui_source = 'https://code.jquery.com/ui/'.$version.'/';
        } else {
            $themes_source = 'plugin://printfriendly/assets/jquery-ui-themes-'.$version.'/themes/';
            $jquery_ui_source = 'plugin://printfriendly/assets/jquery-ui-'.$version.'/';
        }

        $themes = $this->config->get('plugins.printfriendly.libraries.jqueryui_themes');
        if($themes){
            $this->grav['assets']->addCss($themes_source.$themes.'/jquery-ui.css');
        } else {
            $this->grav['assets']->addCss($themes_source.'smoothness/jquery-ui.css');
        }

        if ($this->config->get('plugins.printfriendly.awesome.use_font')) {
            $this->grav['assets']->addCss('plugin://printfriendly/assets/css/font-awesome.min.css');
        }

        $this->grav['assets']->addJs($jquery_ui_source.'jquery-ui.min.js');
    }

    /**
     * Used by the Twig function to generate HTML
     *
     * @param null $route
     * @param array $options
     * @return string
     */
    public function generateLink($route = null, $options = [])
    {
        if ($route === null) {
            return $this->lang->translate('PLUGIN_PF.ERROR');
        }

        $param_sep = $this->grav['config']->get('system.param_sep', ':');

        $pf_bur     = $this->grav['base_url_relative'] . DS;
        $pf_manager = 'pf-manager.json' . DS;
        $pf_task    = 'pftask' . $param_sep . 'pf' . DS;
        $esc_route  = 'route' . $param_sep . str_replace('/', '@', $route) . DS;
        
        $nonce      = Utils::getNonce('pf-form');
        $pf_nonce   = 'pf-nonce' . $param_sep . $nonce;

        $title = '';
        $id = Utils::getNonce($route);
        $btn_id     = "pf-".$id;
        $btn_data   = $pf_bur . $pf_manager . $pf_task . $esc_route . $pf_nonce;

        $print_directly = 'false';
        if( $this->config->get('plugins.printfriendly.basics.print_directly') ){
            $print_directly = 'true';
        }

        $icn_plugin = $this->config->get('plugins.printfriendly.basics.icn_plugin');
        $btn_plugin = $this->config->get('plugins.printfriendly.basics.btn_plugin');
        
        $btn_link = '';
        $btn_link .= ( !empty($icn_plugin) ? '<i class="fa '.$icn_plugin.'" aria-hidden="true"></i>&nbsp;' : '' );
        $btn_link .= ( !empty($btn_plugin) ? $btn_plugin : '' );
        $btn_link = trim($btn_link);

        if( empty($btn_link) ){
            $icn_plugin = 'fa-print';
            $btn_link = '<i class="fa '.$icn_plugin.'" aria-hidden="true"></i>&nbsp;Print';
        }

        $w = $this->config->get('plugins.printfriendly.window.width');
        $h = $this->config->get('plugins.printfriendly.window.height');
        if(!$w) $w = 50;
        if(!$h) $h = 40;
        
        $closeonescape = ( $this->config->get('plugins.printfriendly.window.closeonescape') ? 'true' : 'false' );

        $close_icon = ( $this->config->get('plugins.printfriendly.window.close_icon') ? 'true' : 'false' );

        $btn_confirm = $this->config->get('plugins.printfriendly.window.confirm_button');
        $btn_cancel = $this->config->get('plugins.printfriendly.window.cancel_button');
        if( empty($btn_confirm) ) $btn_confirm = 'Print';
        if( empty($btn_cancel) ) $btn_cancel = 'Close';

        $locator = \Grav\Common\Grav::instance()['locator'];
        $blueprints_path = $locator->findResource('plugins://printfriendly/blueprints.yaml');
        $package_yaml = Yaml::parse(file_get_contents($blueprints_path));
        $version = $package_yaml['version'];

        $html = '
            <div id="dialog-'.$id.'" style="display: none;" title="'.($version ? $version : '').'">
                <div id="print-'.$id.'"></div>
            </div>

            <button id="' . $btn_id . '" class="export" type="button" data-pf="'.$btn_data.'">' . $btn_link . '</button>
            <script>

                function pfPrint( html_content ){
                    $("<iframe>", { name: "pf-frame", class: "pf-frame" }).appendTo("body").contents().find("body").append( html_content );
                    window.frames["pf-frame"].focus();
                    window.frames["pf-frame"].print();
                    setTimeout(() => { $(".pf-frame").remove(); }, 1000);
                };

                $(document).ready(function() {

                    var close_icon = '.$close_icon.';
                    var print_directly = '.$print_directly.';
                    var width = '.$w.';
                    var height = '.$h.';
                    var w = screen.width*width/100;
                    var h = screen.height*height/100
                    w = w.toFixed();
                    h = h.toFixed();
                    $( "#dialog-'.$id.'" ).dialog({
                        width: w,
                        height: h,
                        title: "'.$title.'",
                        autoOpen: false,
                        modal: true,
                        closeOnEscape: '.$closeonescape.',
                        buttons: [
                            {
                                text: "'.$btn_confirm.'",
                                icon: "ui-icon-check",
                                click: function() {
                                    var html_content = $("#print-'.$id.'").html();
                                    pfPrint( html_content );
                                    $("#print-'.$id.'").empty();
                                    $( "div span#fa-'.$id.'" ).remove();
                                    $(this).dialog("close");
                                }
                            },
                            {
                                text: "'.$btn_cancel.'",
                                icon: "ui-icon-closethick",
                                click: function() {
                                    $("#print-'.$id.'").empty();
                                    $( "div span#fa-'.$id.'" ).remove();
                                    $( this ).dialog( "close" );
                                }
                            }
                        ],
                        open: function() {
                            if( !close_icon ) {
                                $(".ui-dialog-titlebar-close").hide();
                            }
                            $( ".ui-dialog-title" ).before( "<span id=\"fa-'.$id.'\"><i class=\"fa '.$icn_plugin.'\" aria-hidden=\"true\"></i>&nbsp;</span>" );
                            $( ".ui-dialog-title" ).css("float","none");
                            $( this ).scrollTop(0);
                        }
                    });


                    $("#'.$btn_id.'").on("click", function () {

                        $("#'.$btn_id.' .'.$icn_plugin.'").removeClass(\''.$icn_plugin.'\').addClass(\'fa-spin fa-spinner\');
                        var element = $("button#'.$btn_id.'");
                        var url = element.attr("data-pf");

                        $.ajax({
                            url: url,
                            dataType: "json",
                            method: "get"
                        }).done(function (data) {
                            if (data.status == "success") {

                                var title = data.title;
                                $("span.ui-dialog-title").text( title );
                                
                                var html_base64_encode = data.html_base64_encode;
                                var decoded = decodeURIComponent(escape(window.atob( html_base64_encode )))
                                $("#print-'.$id.'").html(decoded);
                                
                                if( print_directly ) {
                                    var html_content = $("#print-'.$id.'").html();
                                    pfPrint( html_content );
                                } else {
                                    $("#dialog-'.$id.'").attr("title", "Title");
                                    $("#dialog-'.$id.'").dialog("open");
                                }

                            } else {
                                custom_alert( data.message, data.title );
                            }
                        }).fail(function(jqXHR, textStatus, errorMsg){
                            custom_alert( errorMsg, textStatus );
                        }).always(function(){
                            $("#'.$btn_id.' .fa-spin").removeClass(\'fa-spin fa-spinner\').addClass(\''.$icn_plugin.'\');
                            element.prop( "disabled", false );
                        });
                    });
                });

                function custom_alert( message, title ) {
                    if ( !title ) title = "Alert";
                    if ( !message ) message = "No Message to Display.";
                    $("<div></div>").html( message ).dialog({
                        title: title,
                        resizable: true,
                        modal: true,
                        buttons: {
                            "Ok": function()  {
                                $( this ).dialog( "close" );
                            }
                        }
                    });
                }

            </script>
            ';

        return $html;
    }

}
